File: python_port/src/nls_python/model_generation/inference_rules/exist_elim.py
from __future__ import annotations
from dataclasses import dataclass, field
from typing import Dict, Optional, Set, Tuple

from nls_python.typing_utils import typechecked

from nls_python.fol.formula.exists import Exists
from nls_python.fol.term import Term, SkolemFunction

from ..salient import Salient
from ..tableau_model import TableauModel
from .base import InferenceRule, RuleApplication


@typechecked
class ExistElim(InferenceRule):
    @property
    def branching(self) -> bool:
        return True

    def apply(self, model: TableauModel) -> bool:
        exists_formula = self._find_exist(model)
        if exists_formula is None:
            return False
        application = self._build_application(exists_formula, model)
        if application is None:
            return False
        extensions: Dict[RuleApplication, Set[TableauModel]] = {
            application: {branch for _, branch in application.output_extensions}
        }
        model.add_extensions(extensions)
        model.add_rule_applications({application})
        return True

    def _find_exist(self, model: TableauModel) -> Optional[Exists]:
        for formula in model.get_formulas():
            if isinstance(formula, Exists):
                empty = ExistElimApplication(formula, (), tuple())
                if empty not in model.get_rule_applications():
                    return formula
        return None

    def _build_application(
        self, exists: Exists, model: TableauModel
    ) -> Optional["ExistElimApplication"]:
        branches: Dict[Term, TableauModel] = {}
        individuals = sorted(model.get_individuals(), reverse=True)
        for individual in individuals:
            precondition = exists.apply_precondition(individual.obj)
            if precondition not in model.get_formulas():
                continue
            instantiated = exists.apply(individual.obj)
            branch_model = model.branch_copy()
            branch_model.add_formulas({instantiated})
            branch_model.add_rule_applications({ExistElimApplication(exists, (individual.obj,), tuple())})
            branch_model.set_individuals_salience(
                {Salient(individual.obj, Salient.FULL)}
            )
            branches[individual.obj] = branch_model

        witness_vars = sorted(exists.free_vars(), key=lambda var: var.name)
        witness = SkolemFunction(list(witness_vars))
        witness_formula = exists.apply(witness)
        witness_precondition = exists.apply_precondition(witness)
        witness_model = model.branch_copy()
        witness_model.add_formulas({witness_formula, witness_precondition})
        witness_model.add_individuals({witness}, Salient.FULL)
        witness_model.add_rule_applications({ExistElimApplication(exists, (witness,), tuple())})
        branches[witness] = witness_model

        if not branches:
            return None
        outputs = tuple(branches.items())
        return ExistElimApplication(exists, tuple(branches.keys()), outputs)


@typechecked
@dataclass(frozen=True)
class ExistElimApplication(RuleApplication):
    source: Exists
    initializers: Tuple[Term, ...] = field(compare=False)
    output_extensions: Tuple[Tuple[Term, TableauModel], ...]

    def __hash__(self) -> int:
        return hash(self.source)

    def __eq__(self, other: object) -> bool:
        return isinstance(other, ExistElimApplication) and self.source == other.source

    def describe(self, indentation: int = 0, delim: str = "*") -> str:
        indent = "  " * max(indentation, 0) + delim + " "
        initializers = ",".join(str(i) for i in self.initializers)
        lines = [f"{indent}ExistElim -{initializers}-: {self.source} ->"]
        for term, branch in self.output_extensions:
            branch_header = "  " * (indentation + 1) + delim + f" Witness {term}:"
            lines.append(branch_header)
            lines.append(branch.create_string(indentation + 2, delim))
        return "\n".join(lines)

    def __str__(self) -> str:
        initializers = ",".join(str(i) for i in self.initializers)
        return f"ExistElim -{initializers}-: {self.source}"

File: impl_java/modelGeneration/folModel/rules/ExistElim.java
package modelGeneration.folModel.rules;

import fol.formula.Exists;
import fol.term.SkolemFunction;
import fol.term.Term;
import modelGeneration.Salient;
import modelGeneration.folModel.Model;

import java.util.*;
import java.util.stream.Collectors;

public class ExistElim implements InferenceRule {

    @Override
    public boolean apply(Model model) {
        var e = getExist(model);
        if(e.isEmpty()) return false;
        var maybeApplication = getRuleApplication(e.get(), model);
        if (maybeApplication.isEmpty()) return false;
        var application = maybeApplication.get();
        Set<Model> extensions = application.outputExtensions().values()
                .stream()
                .filter(Optional::isPresent)
                .map(Optional::get)
                .collect(Collectors.toSet());
        model.addExtensions(Map.of(application, extensions));
        model.addRuleApplications(Set.of(application));
        return true;
    }

    @Override
    public boolean isBranching() {
        return true;
    }

    private Optional<Exists> getExist(Model model) {
        return model.getFormulas().stream()
                .filter(Exists.class::isInstance)
                .map(Exists.class::cast)
                .filter(e -> {
                    // Check that the formula wasn't dispatched already
                    var emptyApplication = new ExistsElimRuleApplication(e, Map.of());
                    return !model.getRuleApplications().contains(emptyApplication);
                })
                .findAny();
    }

    private Optional<ExistsElimRuleApplication> getRuleApplication(Exists exists, Model model) {
        var modelRuleApplications = model.getRuleApplications();
        var individuals = model.getIndividuals();
        var emptyApplication = new ExistsElimRuleApplication(exists, Map.of());
        if (modelRuleApplications.contains(emptyApplication))
            return Optional.empty();
        var branches = new HashMap<Term, Optional<Model>>();
        // Create branches for each individual
        for (var individual : individuals.stream().sorted(Comparator.reverseOrder()).toList()) {
            var precondition = exists.applyPrecondition(individual.obj());
            // Skip incomplete preconditions
            if (!model.getFormulas().contains(precondition))
                continue;
            var formula = exists.apply(individual.obj());
            var branchModel = new Model(model.getFormulas(), model.getSentenceDepth(), model.getIndividuals(), model.getExtensions(), model.getRuleApplications(), model.getParent());
            var application = new ExistsElimRuleApplication(exists, Map.of(individual.obj(), Optional.empty()));
            branchModel.addFormulas(Set.of(formula));
            branchModel.addRuleApplications(Set.of(application));
            branchModel.setIndividualsSalience(Set.of(new Salient<>(individual.obj(), Salient.FULL)));
            branches.put(individual.obj(), Optional.of(branchModel));
        }

        // Create new branch with witness
        var witness = new SkolemFunction(exists.freeVars().stream().map(var -> (Term) var).toList());
        var branchFormula = exists.apply(witness);
        var witnessPrecondition = exists.applyPrecondition(witness);
        var witnessModel = new Model(model.getFormulas(), model.getSentenceDepth(), model.getIndividuals(), model.getExtensions(), model.getRuleApplications(), model);
        var application = new ExistsElimRuleApplication(exists, Map.of(witness, Optional.empty()));
        witnessModel.addFormulas(Set.of(branchFormula, witnessPrecondition));
        witnessModel.addIndividuals(Set.of(witness), Salient.FULL);
        witnessModel.addRuleApplications(Set.of(application));
        branches.put(witness, Optional.of(witnessModel));
        return Optional.of(new ExistsElimRuleApplication(exists, branches));
    }

    public record ExistsElimRuleApplication(Exists input,
                                            Map<Term, Optional<Model>> outputExtensions) implements RuleApplication {

        @Override
        public String toString() {
            return getString(0, "");
        }

        @Override
        public boolean equals(Object obj) {
            if (obj instanceof ExistsElimRuleApplication(Exists otherInput, var _)) {
                return this.input.equals(otherInput);
            }
            return false;
        }

        @Override
        public int hashCode() {
            return input.hashCode();
        }

        @Override
        public String getString(int indentation, String delim) {
            if (indentation == 0) {
                StringBuilder sb = new StringBuilder("∃Elim: ");
                sb.append(input).append(" ->");
                if (!outputExtensions.isEmpty())
                    sb.append(" \n");
                for (var entry : outputExtensions.entrySet()) {
                    sb.append("  ").append(entry.getKey()).append(": ");
                    entry.getValue().ifPresent(m -> sb.append(m.createString(2, delim)).append("\n"));
                }
                return sb.toString();
            }
            String outerIndent = "  ".repeat(Math.max(0, indentation)) + delim + " ";
            StringBuilder sb = new StringBuilder(outerIndent)
                    .append("∃Elim : ")
                    .append(input);
            if (!outputExtensions.isEmpty())
                sb.append(" : \n");
            for (var entry : outputExtensions.entrySet()) {
                sb.append("  ".repeat(indentation + 1)).append(delim).append(" ").append(entry.getKey()).append(": ");
                entry.getValue().ifPresent(m -> sb.append(m.createString(indentation + 2, delim.repeat(2))).append("\n"));
            }
            return sb.toString();
        }
    }
}
